src/collections/entry_map.rs

```diff
@@ -0,0 +1,288 @@
+use super::map::Iter as MapIter;
+use super::map::Map;
+use crate::encoding::{Decode, Encode};
+use crate::store::DefaultBackingStore;
+use std::hash::Hash;
+use std::ops::RangeBounds;
+
+use super::{Entry, Next};
+use crate::state::*;
+use crate::store::*;
+use ed::*;
+
+pub struct EntryMap<T: Entry, S = DefaultBackingStore> {
+    map: Map<T::Key, T::Value, S>,
+}
+
+impl<T: Entry, S> From<EntryMap<T, S>> for () {
+    fn from(_map: EntryMap<T, S>) {}
+}
+
+impl<T: Entry, S> State<S> for EntryMap<T, S>
+where
+    T::Key: Encode + Terminated + Eq + Hash + Ord,
+    T::Value: State<S>,
+{
+    type Encoding = ();
+
+    fn create(store: Store<S>, _: ()) -> Result<Self>
+    where
+        S: Read,
+    {
+        Ok(EntryMap {
+            map: Map::create(store, ())?,
+        })
+    }
+
+    fn flush(self) -> Result<()>
+    where
+        S: Write,
+    {
+        self.map.flush()
+    }
+}
+
+impl<T, S> EntryMap<T, S>
+where
+    T: Entry,
+    T::Key: Encode + Terminated + Eq + Hash + Ord + Copy,
+    T::Value: State<S>,
+    S: Read,
+{
+    pub fn insert(&mut self, entry: T) -> Result<()> {
+        let (key, value) = entry.into_entry();
+        self.map.entry(key)?.or_insert(value.into())?;
+
+        Ok(())
+    }
+
+    pub fn delete(&mut self, entry: T) -> Result<()> {
+        let (key, _) = entry.into_entry();
+        self.map.remove(key)?;
+
+        Ok(())
+    }
+
+    pub fn contains(&self, entry: T) -> Result<bool> {
```

Comment:
Refactored the original contains to contains_entry_key and added a contains_key method. Some trait bound issues required that I manually check returned values rather than the entire entry. Only required Eq on T::Value rather than T.